// @expo/config-plugins/build/utils/warnings.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.hasWarningsIOS = hasWarningsIOS;
exports.hasWarningsAndroid = hasWarningsAndroid;
exports.addWarningAndroid = addWarningAndroid;
exports.addWarningIOS = addWarningIOS;
exports.addWarningForPlatform = addWarningForPlatform;
exports.flushWarningsAndroid = flushWarningsAndroid;
exports.flushWarningsIOS = flushWarningsIOS;
let _warningsIOS = [];
let _warningsAndroid = [];

function hasWarningsIOS() {
  return !!_warningsIOS.length;
}

function hasWarningsAndroid() {
  return !!_warningsAndroid.length;
}

function addWarningAndroid(tag, text, link) {
  _warningsAndroid = [..._warningsAndroid, [tag, text, link]];
}

function addWarningIOS(tag, text, link) {
  _warningsIOS = [..._warningsIOS, [tag, text, link]];
}

function addWarningForPlatform(platform, tag, text, link) {
  if (platform === 'ios') {
    addWarningIOS(tag, text, link);
  } else {
    addWarningAndroid(tag, text, link);
  }
}

function flushWarningsAndroid() {
  const result = _warningsAndroid;
  _warningsAndroid = [];
  return result;
}

function flushWarningsIOS() {
  const result = _warningsIOS;
  _warningsIOS = [];
  return result;
}
